payload/builder/payload.go

```diff
@@ -164,10 +164,27 @@ func (pb *PayloadBuilder) RetrievePayload(
 		return nil, ErrPayloadBuilderDisabled
 	}
 
-	// Attempt to see if we previously fired off a payload built for
-	// this particular slot and parent block root.
+	// With optimistic block building enabled, multiple payloads can be available
+	// when it's the proposer turn to build. We select them as follows:
+	// - First we check if node has already built a payload for the requested slot. If so we use that
+	// payload; note that such a payload may not be available (no payloadID associated or payload is stale).
+	// - Secondly we try and reuse the latest payload we verified, even if produced by other validators.
+	// The reason we have to do this has to do with an ENGINE API invariant: the node
+	// could be asked to build a block at slot H even if it already verified a block at that slot H.
+	// This happens if the block passes verification but is not finalized (e.g. due to network issue).
+	// In such a case, the EVM may not be able to serve a new payload (e.g. if it has received a
+	// FCU call with HEAD == H+1). To avoiding a failure in building the block
```

Comment:
nit: technically the EL is unable to serve a new payload because the FCU call contains (FINALIZED == H). The HEAD == H+1 does not actually matter.